pkg/apiserver/resthandler.go

```diff
@@ -446,6 +418,128 @@ func PatchResource(r rest.Patcher, scope RequestScope, typer runtime.ObjectTyper
+		return nil, err
+	}
+
+	return finishRequest(timeout, func() (runtime.Object, error) {
+		// update should never create as previous get would fail
+		updateObject, _, updateErr := r.Update(ctx, objToUpdate)
+		if errors.IsConflict(updateErr) {
+
+			// on a conflict,
+			// 1. build a strategic merge patch from originalJS and the patchedJS
+			// 2. build a strategic merge patch from originalJS and the currentJS
+			// 3. ensure no conflicts between the two patches
+			// 4. apply the #1 patch to the currentJS object
+			// 5. retry the update
+			currentObject, err := r.Get(ctx, name)
+			if err != nil {
+				return nil, err
+			}
+			currentObjectJS, err := codec.Encode(currentObject)
+			if err != nil {
+				return nil, err
+			}
+
+			currentPatch, err := strategicpatch.CreateStrategicMergePatch(originalObjJS, currentObjectJS, r.New())
+			if err != nil {
+				return nil, err
+			}
+			originalPatch, err := strategicpatch.CreateStrategicMergePatch(originalObjJS, originalPatchedObjJS, r.New())
+			if err != nil {
+				return nil, err
+			}
+
+			diff1 := make(map[string]interface{})
+			if err := json.Unmarshal(originalPatch, &diff1); err != nil {
+				return nil, err
+			}
+			diff2 := make(map[string]interface{})
+			if err := json.Unmarshal(currentPatch, &diff2); err != nil {
+				return nil, err
+			}
+			if hasConflicts(diff1, diff2) {
+				return updateObject, updateErr
+			}
+
+			newlyPatchedObjJS, err := getPatchedJS(api.StrategicMergePatchType, currentObjectJS, originalPatch, versionedObj)
+			if err != nil {
+				return nil, err
+			}
+			if err := codec.DecodeInto(newlyPatchedObjJS, objToUpdate); err != nil {
+				return nil, err
+			}
+
+			updateObject, _, updateErr = r.Update(ctx, objToUpdate)
+		}
+
+		return updateObject, updateErr
+	})
+}
+
+// hasConflicts returns true if the left and right JSON interface objects overlap with
+// different values in any key.  The code will panic if an unrecognized type is passed
+// (anything not returned by a JSON decode).  All keys are required to be strings.
+func hasConflicts(left, right interface{}) bool {
+	switch typedLeft := left.(type) {
+	case map[string]interface{}:
+		switch typedRight := right.(type) {
+		case map[string]interface{}:
+			for key, leftValue := range typedLeft {
+				if rightValue, ok := typedRight[key]; ok && hasConflicts(leftValue, rightValue) {
+					return true
+				}
+			}
+			return false
+		default:
+			return true
+		}
+	case []interface{}:
+		switch typedRight := right.(type) {
+		case []interface{}:
+			if len(typedLeft) != len(typedRight) {
+				return true
+			}
+			for i := range typedLeft {
+				if hasConflicts(typedLeft[i], typedRight[i]) {
+					return true
+				}
+			}
+			return false
+		default:
+			return true
+		}
+	case string, float64, bool, int, int64, nil:
+		return !reflect.DeepEqual(left, right)
+	default:
+		panic(fmt.Sprintf("unknown type: %v", reflect.TypeOf(left)))
```

Comment:
> why panic? Why not return error?

I've refactored to a common method (that was a naughty copy/paste) and replaced the panic with an error.